### PR TITLE
Automated release notes and process

### PR DESCRIPTION
**Pull request by devinmatte**
Automating release notes on new tags

Using https://github.com/marketplace/actions/release-changelog-builder (examples visible there)

# Process

1. Add a new release commit bumping to the version in package.json
2. Add a new tag tied to that version bump commit
3. Push that commit and tag to `master`
4. Profit (CI runs on tag and creates new release)

Alternatively, included here is a bash script that automates most of it (except the last step for safety)

In order for this to work, we would also need PRs tagged with relevant labels (see config), we can automate some if we want

### CODE REVIEW

File: release.bash [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+# Create new commit
+git add package.json package-lock.json
+git commit -m "Bump to $1"
+
+# Add a tag
+git tag $1
+
+printf "\n\n🚇🚇🚇 New Release Ready 🚇🚇🚇\n\n"
+printf "The release has been created and tagged. Run the below command with write permissions to master\n\n"
+printf "    git push origin master && git push origin $1\n"
```
Comment (thiteixeira) [inline release.bash:22]: Tip for readability: Use [zx](https://github.com/google/zx)

If you are not familiar with zx, it lets you write bash scripts in plain javascript 🙂

File: .github/workflows/release.yml [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+  # It is also possible to trigger via the API:
+  # https://docs.github.com/en/rest/actions/workflows#create-a-workflow-dispatch-event
+  workflow_dispatch:
+    inputs:
+      tag:
+        description: "The tag of the version that is being deployed"
+        required: true
+        type: string
+
+name: Make a GitHub release with a deployment changelog
```
Comment (thiteixeira) [inline .github/workflows/release.yml:12]: Can we move `name` to the first line? I'm not sure if there's a best practice, but it just makes it easier to read and more organized

**Review by thiteixeira — COMMENTED** [review]
Overall it looks good. Do you have an example workflow run?

File: .github/release-config.json [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+        "title": "### 🚆 CI/CD",
+        "labels": ["ci/cd"]
+      }
+    ],
+    "ignore_labels": ["release"],
+    "pr_template": "- ${{TITLE}} (#${{NUMBER}})",
+    "tag_resolver": {
+      "method": "sort"
+    },
+    "base_branches": ["master"]
```
[the remark this answers is not in the document]
Comment (thiteixeira) [reply .github/release-config.json:25]: The argument here is that we eventually are going to rename `master` to `main`, so having both will prevent the workflow from breaking in the future

File: .github/workflows/release.yml [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+          script: |
+            const release = await github.rest.repos.getLatestRelease({
+              owner: context.repo.owner,
+              repo: context.repo.repo
+            })
+            return release.data.tag_name
+
+      - name: Build Release Changelog
+        id: github_release
+        uses: mikepenz/release-changelog-builder-action@v3
```
[the remark this answers is not in the document]
Comment (devinmatte) [reply .github/workflows/release.yml:34]: I thought GitHub only verified businesses and not individuals

File: .github/workflows/release.yml [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+        env:
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+
+      - name: Create Release
+        id: create_release
+        uses: actions/create-release@v1
+        env:
+          # Default token (secrets.GITHUB_TOKEN) is sufficient
+          # because this step does not require access to another private repo
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
```
[the remark this answers is not in the document]
Comment (devinmatte) [reply .github/workflows/release.yml:49]: I'm pretty sure this exists by default

**Comment by devinmatte** [comment]
In order for this to run it needs to be in `master` with a previous tag to compare to, in a repo with PRs. I can try and imitate it in my fork, but the best example of how it works is in https://github.com/mikepenz/release-changelog-builder-action/releases itself

File: .github/workflows/release.yml (hunk 4 above)
[the remark this answers is not in the document]
Comment (devinmatte) [reply .github/workflows/release.yml:34]: On the safety/reliability of this one, It's used by nearly 2k repos https://github.com/mikepenz/release-changelog-builder-action/network/dependents?package_id=UGFja2FnZS0yOTQyNTU4ODY5

[the remark this answers is not in the document]
Comment (thiteixeira) [reply .github/workflows/release.yml:34]: We are already using another one [here](https://github.com/transitmatters/new-train-tracker/blob/master/.github/workflows/deploy.yml#L15)

**Review by nathan-weinberg — COMMENTED** [review]
Thanks for opening this @devinmatte I really like the look of the workflow. I think we should definitely discuss this one in the meeting as we'll want to establish some common practices (probably codified in the Google Drive) with the usage of it.

Could you add a badge to the README for the new workflow after `deploy`?

File: .github/workflows/release.yml (hunk 4 above)
[the remark this answers is not in the document]
Comment (nathan-weinberg) [reply .github/workflows/release.yml:34]: Right now, we allow actions from any source as an org policy
image

**Pull request closed without merging**